doc-util/render.libsonnet

```diff
@@ -0,0 +1,324 @@
+      ])
+      + (root.templates.sectionLink % {
+           abbr: section.type.abbr,
+           linkName: section.linkName,
+           link:
+             std.asciiLower(
+               std.strReplace(
+                 std.strReplace(root.renderSectionTitle(section, prefixes), '.', '')
+                 , ' ', '-'
+               )
+             ),
+         })
+      + (
+        if std.length(section.subSections) > 0
+        then '\n' + root.index(
+          section.subSections,
+          depth + 1,
+          prefixes + [section.name]
+        )
+        else ''
+      )
+      for section in sections
+    ]),
+
+  sections: {
+    base: {
+      subSections: [],
+      values: [],
+    },
+    object(key, doc, obj, depth):: self.base {
+      name: std.strReplace(key, '#', ''),
+
+      local processed = root.process(obj, depth=depth + 1),
+
+      subSections: processed.sections,
+
+      values: processed.values,
+
+      type: { full: 'object', abbr: 'obj' },
+
+      abbr: self.type.abbr,
+
+      doc:
+        if self.type.full in doc
+        then doc[self.type.full]
+        else { help: '' },
+
+      help: self.doc.help,
+
+      linkName: self.name,
+
+      content:
+        if self.help != ''
+        then self.help + '\n'
+        else '',
+    },
+
+    'function'(key, doc):: self.base {
+      name: std.strReplace(key, '#', ''),
+
+      type: { full: 'function', abbr: 'fn' },
+
+      abbr: self.type.abbr,
+
+      doc: doc[self.type.full],
+
+      help: self.doc.help,
+
+      args: std.join(', ', [
+        if arg.default != null
+        then arg.name + '=' + arg.default
+        else arg.name
+        for arg in self.doc.args
+      ]),
+
+      linkName: '%(name)s(%(args)s)' % self,
+
+      content: '```ts\n%(name)s(%(args)s)\n```\n\n%(help)s' % self,
+
+    },
+
+    value(key, doc, obj):: self.base {
+      name: std.strReplace(key, '#', ''),
+      type: doc.value.type,
+      help: doc.value.help,
+      value: obj,
+    },
+  },
+
+  process(obj, filename='', depth=0)::
+    std.foldl(
+      function(acc, key)
+        acc +
+        // Package definition
+        if key == '#'
+        then obj[key] { filename: filename }
+
+        // Field definition
+        else if std.startsWith(key, '#')
+        then (
+          local realKey = key[1:];
+          if 'value' in obj[key]
+          then {
+            values+: [root.sections.value(
+              key,
+              obj[key],
+              obj[realKey]
+            )],
+          }
+          else if 'function' in obj[key]
+          then {
+            functionSections+: [root.sections['function'](
+              key,
+              obj[key],
+            )],
+          }
+          else if 'object' in obj[key]
+          then {
+            objectSections+: [root.sections.object(
+              key,
+              obj[key],
+              obj[realKey],
+              depth
+            )],
+          }
+          else {}
```

Comment:
should this not be an error, reaching here?

Reply:
Thing can be undocumented, which I think is reasonable. This is also forgiven by the current renderer.

Reply:
okay, makes sense.